library/src/scala/IArray.scala

```diff
@@ -102,7 +100,7 @@ object opaques:
 
   /** Flattens a two-dimensional array by concatenating all its rows
     * into a single array. */
-  extension [T](arr: IArray[T]) def flatten[U: ClassTag](using T => Iterable[U]): IArray[U] =
+  extension [T](arr: IArray[T]) def flatten[U: ClassTag](using asIterable: T => Iterable[U]): IArray[U] =
```

Comment:
If we are against adding new implicit conversions should we simplify the definitions of `flatMap`, `WithFilter.flatMap`, `flatten`, `unzip`, `unzip3`, `transpose`?

For this one the alternative is to define it as
```scala
 extension [T](arr: IArray[Iterable[T]]) def flatten: IArray[T] = ...
```
We could also add an overload for 
```scala
 extension [T](arr: IArray[IArray[T]]) def flatten: IArray[T] = ...
```
@odersky @sjrd WDYT?

Reply:
We'll need an implicit conversion from `IArray[T]` to `Iterable[T]` anyway, because I can to be able to give `IArray`s in the `flatMap` methods of `List` and other collections.

Reply:
Should the conversion be to `Iterable[T]` or to `immutable.ArraySeq[T]` directly?

Reply:
`immutable.ArraySeq` sounds reasonable to me.

Reply:
This allowed me to remove many redundant operations. See the second commit.

Reply:
The problem with using `Conversion` instances is that you get a feature warning. 
```
there were 1 feature warning(s); re-run with -feature for details
```
If you run with `-feature` you get something like this
```
  |Use of implicit conversion given instance genericWrapAnyRefIArray in object IArray should be enabled
  |by adding the import clause 'import scala.language.implicitConversions'
  |or by setting the compiler option -language:implicitConversions.
  |See the Scala docs for value scala.language.implicitConversions for a discussion
  |why the feature should be explicitly enabled.
```
To prevent this you can use an old-style implicit def. Or, alternatively, change `checkImplicitConversionUseOK` to make a special case for conversions coming from `IArray`. But the second alternative requires a discussion what the policy wrt implicit conversions should be. So I think it's easiest for now to simply revert to implicit defs for the conversions here.

Reply:
You mean `flatMap` as in the following example?
```scala
for
  x <- List(1, 2)
  y <- IArray(1, 2)
yield ...
```
For this, we do not need the implicit conversion.

Reply:
I used `implicit def`s and aligned the implementation with the one for Arrays